### DISCUSSION TITLE
404 error persists after trying all fixes

### DISCUSSION THREAD

**Discussion by davesimms44** [Project-MONAI/MONAILabel#1417] [Q&A]
Hello all,

I am having issues with MONAIBundle and the Github API - 
401 Client Error: Unauthorized for url: https://api.github.com/repos/Project-MONAI/model-zoo/releases
also 
 404 Client Error: Not Found for url: https://api.github.com/repos/Project-MONAI%5Cmodel-zoo/releases

I have reinstalled using GIthub (rather than PyPi) as per this[ thread ](https://github.com/Project-MONAI/MONAILabel/discussions/1297). 
In that thread, [fix 5775](https://github.com/Project-MONAI/MONAI/pull/5787) is linked. This seems to suggest there is an error with the URL (that was fixed?), and i am unsure how to fix since it still wont work for me.

I can get access once, but the bundle will not work, and returns this error : 

[2023-04-23 14:25:22,349] [12260] [MainThread] [ERROR] (uvicorn.error:119) - Traceback (most recent call last):
  File "C:\Users\dave_\anaconda3\envs\MonaiGit\lib\site-packages\starlette\routing.py", line 635, in lifespan
    async with self.lifespan_context(app):
  File "C:\Users\dave_\anaconda3\envs\MonaiGit\lib\site-packages\starlette\routing.py", line 530, in __aenter__
    await self._router.startup()
  File "C:\Users\dave_\anaconda3\envs\MonaiGit\lib\site-packages\starlette\routing.py", line 612, in startup
    await handler()
  File "C:\Users\dave_\anaconda3\envs\MonaiGit\lib\site-packages\monailabel\app.py", line 106, in startup_event
    instance = app_instance()
  File "C:\Users\dave_\anaconda3\envs\MonaiGit\lib\site-packages\monailabel\interfaces\utils\app.py", line 51, in app_instance
    app = c(app_dir=app_dir, studies=studies, conf=conf)
  File "C:\Users\dave_\apps\monaibundle\main.py", line 36, in __init__
    self.models = get_bundle_models(app_dir, conf)
  File "C:\Users\dave_\anaconda3\envs\MonaiGit\lib\site-packages\monailabel\utils\others\generic.py", line 274, in get_bundle_models
    exit(-1)
  File "C:\Users\dave_\anaconda3\envs\MonaiGit\lib\_sitebuiltins.py", line 26, in __call__
    raise SystemExit(code)
SystemExit: -1
[2023-04-23 14:25:22,349] [12260] [MainThread] [ERROR] (uvicorn.error:56) - Application startup failed. Exiting.

Once I get that error, I am unable to access the API to reproduce it unless I remove my auth token and reset that completely. Once I :struggle to reset the key and set it again, I can reproduce the above  error once  with this command or a similar one:
monailabel start_server --app monaibundle --studies Task09_Spleen/imagesTr --conf models swin_unetr_btcv_segmentation 

 than the very next start_server command loops me back to 404 errors.

I have scoured all available resources on Git,Slicer discourse, and docs, and can not find a fix for these issues. MONAILabel works fine.
Any help is appreciated.

Thanks, Dave

**Comment by davesimms44** [comment]
I have tried with different models from the zoo as well, from -c models list

**Comment by tangy5** [comment]
HI @davesimms44 , if you are using the released version of MONAI Label, such as 0.6.0.
You could start using monaibundle in these steps, use swin_unetr_btcv_segmentation as an example:

`monailabel apps --name monaibundle --download --output apps
`

then 
`monailabel start_server --app apps/monaibundle --studies datasets/Task09_Spleen/imagesTr --conf models swin_unetr_btcv_segmentation --conf auth_token <Github token>
`
Make sure the auth_token is added here, the Model_Zoo has a rate limit, and we need to provide the author token to avoid 401 complains.

**Reply by davesimms44** [reply]
Thanks for the reply. 
Unfortunately, I have tried passing --conf auth_token with the key and it results in the same error after the first successful attempt.

**Reply by davesimms44** [reply]
In fact, I think passing that command may be the only way it works the one time . Regardless, it ceases to work after the first one unless I start over, delete the keys, and set them again. Even then, it takes some tinkering until I can get it to work once.

**Reply by tangy5** [reply]
The error you post above seems a "model do not exist error", can you see the "swin_unetr_btcv_segmentation" model is downloading when you start monailabel server?

**Reply by davesimms44** [reply]
I have tried with models that do exist in the zoo as well. lung_nodule_ct_detection etc.

**Reply by davesimms44** [reply]
That swin_unet command is from https://docs.monai.io/projects/label/en/latest/quickstart.htmls which seems to be outdated.

**Reply by tangy5** [reply]
Thanks. I have tried my side, but it all looks fine. I'm wondering if it's a Windows related issue, can you post reproducing command snippets, so that we can reproduce? Thank you again.

**Reply by davesimms44** [reply]
Yes, certainly. I will create a fresh conda env and go thru all the steps and post the results.

 From what I can gather, yes it is a windows issue (see the links in the initial post )

If I cant get this to work, 
I am going to try the fix in this [thread](https://discourse.slicer.org/t/monai-error-cannot-import-name/28222/3),  setting the keys to local , and downloading the .zip,

**Reply by tangy5** [reply]
Thanks, I saw the error logs in the thread, it looks like a MONAI bundle API's issue on Windows OS. Once we can reproduce it, we can ask the right person to investigate it. Thanks again for raising the issue.

**Reply by tangy5** [reply]
Just tried on Windows, I think I can reproduce the 404 error. We need to report this to the MONAI Bundle folks. Let's create an issue there. Will update us.

**Comment by davesimms44** [comment]
Thanks so much. I had actually meant to post it in the bundle discussion
😅.
Shoukd I create an issue or will you?

Dave

**Reply by tangy5** [reply]
I can reproduce the error in Windows by using this code snippet:

```
import monai
from monai.bundle import download
download(name="swin_unetr_btcv_segmentation")
```

If this also occurs on your side, can you create an issue on this page:
https://github.com/Project-MONAI/MONAI/issues

I will follow up, I think the bundle API creator doesn't test and verify on Windows, but we have lots of users with Windows machines. 

Thanks again.

**Comment by SachidanandAlle** [comment]
Its fixed in latest rc version of monai.. so in that case you should either upgrade monai version or monailabel..
https://pypi.org/project/monailabel/0.7.0rc5/

or let us know which versions of monai and monailabel are installed in your system

And is this problem happening on one bundle or for any bundle?

**Reply by tangy5** [reply]
@davesimms44 , yes, looks like there is a fix already. for a direct try, you can uninstall monai and reinstall with "pip install monai==1.2.0rc4"

**Comment by davesimms44** [comment]
Yes, I had thought it had been fixed! I am in fact using 0.6 in this
environment.
Will report back.
Thanks immensely.

**Comment by davesimms44** [comment]
0.7.0rc5 fixed the problem. Command from the Quickstart guide was successful and the swin_unet model is downloading now. 
To confirm, as I am writing a how-to blog for clinicians - **the MONAI_ZOO_AUTH_TOKEN still has to be set for access to monaibundle?**
 I do not want to make them create a token if not necessary.

Thanks,

Dave

**Comment by tangy5** [comment]
If you are using lastest MONAI Label rc version, the token is optional, you don't need to provide it until you encounter a rate limit. 

Thanks for sharing it. Hope the workflow can help clinicians. 

In addition, we have a series of tutorials for beginners who'd like to try monailabel step by step.

https://github.com/Project-MONAI/tutorials/blob/main/monailabel/monailabel_monaibundle_3dslicer_multiorgan_seg.ipynb

**Comment by davesimms44** [comment]
Perfect. Thanks so much.
